src/lib/reasoners/satml_frontend.ml

```diff
@@ -815,575 +827,588 @@ module Make (Th : Theory.S) : Sat_solver_sig.S = struct
+              env.last_saved_model <- None;
+              env.last_saved_objectives <- None;
+              env.inst <- inst;
+              env.guards.current_guard <- b;
+              let declare_top =
+                try
+                  Stack.pop env.declare_tail
+                with Stack.Empty ->
+                  Errors.error (Run_error Stack_underflow)
+              in
+              env.declare_top <- declare_top;
+            )
+          ~max:to_pop
+          ~elt:()
+          ~init:()
+
+      let add_guard env gf =
+        let current_guard = env.guards.current_guard in
+        {gf with E.ff = E.mk_imp current_guard gf.E.ff}
+
+      let unsat env gf =
+        checks_implemented_features ();
+        let gf = add_guard env gf in
+        Debug.unsat gf;
+        (*fprintf fmt "FF.unsat@.";*)
+        (* In dfs_sat goals' terms are added to env.inst *)
+        env.inst <-
+          Inst.add_terms env.inst
+            (E.max_ground_terms_rec_of_form gf.E.ff) gf;
+        try
+          assert (SAT.decision_level env.satml == 0);
+          let _updated = assume_aux ~dec_lvl:0 env [gf] in
+          let max_t = max_term_depth_in_sat env in
+          env.inst <- Inst.register_max_term_depth env.inst max_t;
+          unsat_rec_prem env ~first_call:true;
+          assert false
+        with
+        | IUnsat (_env, dep) ->
+          assert begin
+            Ex.fold_atoms
+              (fun e b -> match e with
+                 | Ex.Bj _ -> false (* only used in fun_sat *)
+                 | Ex.Literal _ | Ex.Fresh _ -> false (* bug if this happens *)
+                 | Ex.RootDep _ | Ex.Dep _ -> b
+              )dep true
+          end;
+          dep
+
+      let assume env gf _dep =
+        (* dep currently not used. No unsat-cores in satML yet *)
+        assert (SAT.decision_level env.satml == 0);
+        try ignore (assume_aux ~dec_lvl:0 env [add_guard env gf])
+        with | IUnsat (_env, dep) -> raise (Unsat dep)
+             | Util.Timeout ->
+               (* don't attempt to compute a model if timeout before
+                  calling unsat function *)
+               i_dont_know env (Timeout Assume)
+             | Util.Step_limit_reached n ->
+               (* When reaching the step limit on an assume, we do not want to
+                  answer 'unknown' right away. *)
+               env.unknown_reason <- Some (Step_limit n)
+
+      (* instrumentation of relevant exported functions for profiling *)
+      let assume t ff dep =
+        if not (Options.get_timers ()) then assume t ff dep
+        else
+          try
+            Timers.exec_timer_start Timers.M_Sat Timers.F_assume;
+            assume t ff dep;
+            Timers.exec_timer_pause Timers.M_Sat Timers.F_assume;
+          with exn ->
+            Timers.exec_timer_pause Timers.M_Sat Timers.F_assume;
+            raise exn
+
+      (* HOTFIX: we can assert new formula in [env.satml] only at the level of
```

Comment:
Needs a reference to a GH issue.

Reply:
Also see previous comment:

> That is unfortunate we have to do this even if we never call (get-value); is there a chance we could call cancel_until at the moment (get-value) is called instead?

Reply:
I fix it by adding a new function `SAT.reset_decisions` in the SAT API. Now we can reset all the decisions after doing a `check-sat` and saved the boolean model just before.